### PR TITLE
Implements timeout for the Handshake miniprotocol

### PR DESCRIPTION
**Pull request by karknu**

### CODE REVIEW

File: ouroboros-network-framework/src/Ouroboros/Network/Socket.hs [hunk 1]
```diff
@@ -236,21 +227,27 @@ connectToNode' sn versionDataCodec NetworkConnectTracers {nctMuxTracer, nctHands
...
-         Right app -> do
-            traceWith muxTracer $ Mx.MuxTraceHandshakeClientEnd (diffTime ts_end ts_start)
-            Mx.muxStart muxTracer (toApplication (app connectionId)) bearer
+
+    case mmapp of
+         Left (errDrv :: ProtocolLimitFailure)-> do
+            traceWith muxTracer $ Mx.MuxTraceHandshakeClientError errDrv (diffTime ts_end ts_start)
+            throwIO errDrv
+         Right r ->
+             case r of
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/Socket.hs]: why not inline the two pattern matches, e.g.
```
Right (Left err) -> 
Right (Right app)  ->
```

File: ouroboros-network-framework/src/Ouroboros/Network/Socket.hs [hunk 2]
```diff
@@ -344,22 +341,27 @@ beginConnection sn muxTracer handshakeTracer versionDataCodec acceptVersion fn t
...
-          Right mkapp ->
-            case mkapp peerid of
-              SomeResponderApplication app -> do
-                traceWith muxTracer' $ Mx.MuxTraceHandshakeServerEnd
-                Mx.muxStart muxTracer' (toApplication app) bearer
+        case mmapp of
+             Left (errDrv :: ProtocolLimitFailure) -> do
+               traceWith muxTracer' $ Mx.MuxTraceHandshakeServerError errDrv
+               throwIO errDrv
+             Right mapp  ->
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/Socket.hs]: The same here :).

File: ouroboros-network-framework/test/Test/Ouroboros/Network/Driver.hs [hunk 3]
```diff
@@ -1,10 +1,12 @@
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE KindSignatures      #-}
 {-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE PolyKinds           #-}
```
Comment (coot) [inline ouroboros-network-framework/test/Test/Ouroboros/Network/Driver.hs]: either align all of `#-}` or none.

Comment (coot) [reply ouroboros-network-framework/test/Test/Ouroboros/Network/Driver.hs]: I am actually surprised we need additional extensions, e.g. `PolyKinds`; the removed `runPeerWithByteLimit` didn't seem to require it in this module.

Comment (karknu) [reply ouroboros-network-framework/test/Test/Ouroboros/Network/Driver.hs]: The limits with per protocol state types are more complex than the previous byte limit, see `byteLimitsReqResp` bellow.

**Comment by coot** [comment]
Related to #1395

**Comment by coot** [comment]
Related to #1575

**Pull request merged**